gio/src/task.rs

```diff
@@ -382,6 +384,64 @@ impl<V: ValueType + Send> Task<V> {
 unsafe impl<V: ValueType + Send> Send for Task<V> {}
 unsafe impl<V: ValueType + Send> Sync for Task<V> {}
 
+#[derive(Debug)]
+pub struct JoinHandle<T: Send + 'static> {
+    rx: oneshot::Receiver<T>,
+    cancellable: Option<Cancellable>,
+}
+
+impl<T: Send + 'static> JoinHandle<T> {
+    #[inline]
+    fn new(cancellable: Option<Cancellable>) -> (Self, oneshot::Sender<T>) {
+        let (tx, rx) = oneshot::channel();
+        (Self { rx, cancellable }, tx)
+    }
+    #[inline]
+    pub fn cancellable(&self) -> Option<&Cancellable> {
+        self.cancellable.as_ref()
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum JoinError {
+    #[error("task cancelled")]
+    Cancelled,
```

Comment:
As discussed on Matrix, it would be great if this would catch panics like tokio also does and returns them as another kind of error instead of aborting the process.

Similarly, the normal `spawn()` should probably do the same and also return such a join handle.